code/modules/projectiles/projectile.dm

```diff
@@ -163,25 +163,42 @@
 		if(forcedodge)
 			loc = target_turf
 		return FALSE
-	var/permutation = A.bullet_act(src, def_zone) // searches for return value, could be deleted after run so check A isn't null
+	var/permutation = select_target(A,target_turf,forcedodge) // searches for return value, could be deleted after run so check A isn't null
 	if(permutation == -1 || forcedodge)// the bullet passes through a dense object!
 		loc = target_turf
 		if(A)
 			permutated.Add(A)
 		return FALSE
-	else
-		if(A && A.density && !ismob(A) && !(A.flags_1 & ON_BORDER_1)) //if we hit a dense non-border obj or dense turf then we also hit one of the mobs on that tile.
-			var/list/mobs_list = list()
-			for(var/mob/living/L in target_turf)
-				mobs_list += L
-			if(mobs_list.len)
-				var/mob/living/picked_mob = pick(mobs_list)
-				if(!prehit(picked_mob))
-					return FALSE
-				picked_mob.bullet_act(src, def_zone)
 	qdel(src)
 	return TRUE
 
+/obj/item/projectile/proc/select_target(atom/A,target_turf)
+	if((A && A.density && !(A.flags_1 & ON_BORDER_1)) && (istype(A,/obj/machinery) || isturf(A))) //if we hit a dense non-border obj or dense turf then we also hit one of the mobs on that tile.
+		var/list/mobs_list = list()
+		var/list/machine_list = list()
+		for(var/mob/living/L in target_turf)
+			mobs_list += L
+		for(var/obj/machinery/m in target_turf)

```

Comment:
why can machines be hit but not structures?

Reply:
mobs/machines in structures will already be hit before the structure is.
If there are no mobs.machines then the bullet will act on the structure at the last else.